tests/common_coin.rs

```diff
@@ -0,0 +1,130 @@
+//! Common Coin tests
+
+extern crate env_logger;
+extern crate hbbft;
+#[macro_use]
+extern crate log;
+extern crate pairing;
+extern crate rand;
+
+mod network;
+
+use std::iter::once;
+
+use rand::Rng;
+
+use hbbft::common_coin::CommonCoin;
+
+use network::{Adversary, MessageScheduler, NodeUid, SilentAdversary, TestNetwork, TestNode};
+
+/// Tests a network of Common Coin instances with an optional expected value. Outputs the computed
+/// common coin value if the test is successful.
+fn test_common_coin<A>(
+    mut network: TestNetwork<A, CommonCoin<NodeUid, String>>,
+    expected_coin: Option<bool>,
+) -> bool
+where
+    A: Adversary<CommonCoin<NodeUid, String>>,
+{
+    let ids: Vec<NodeUid> = network.nodes.keys().cloned().collect();
+    for id in ids {
+        network.input(id, ());
+    }
+
+    // Handle messages until all good nodes have terminated.
+    while !network.nodes.values().all(TestNode::terminated) {
+        network.step();
+    }
+    let mut expected = expected_coin;
+    // Verify that all instances output the same value.
+    for node in network.nodes.values() {
+        if let Some(b) = expected {
+            assert!(once(&b).eq(node.outputs()));
+        } else {
+            assert_eq!(1, node.outputs().len());
+            expected = Some(node.outputs()[0]);
+        }
+    }
+    // Now `expected` is the unique output of all good nodes.
+    expected.unwrap()
+}
+
+const GOOD_SAMPLE_SET: f64 = 400.0;
+
+/// The count of throws of each side of the coin should be approaching 50% with a sufficiently large
+/// sample set. This check assumes logarithmic growth of the expected number of throws of one coin
+/// size.
+fn check_coin_distribution(num_samples: usize, count_true: usize, count_false: usize) {
+    const EXPECTED_SHARE: f64 = 0.48;
+    let max_gain = GOOD_SAMPLE_SET.log2();
+    let num_samples_f64 = num_samples as f64;
+    let gain = num_samples_f64.log2().min(max_gain);
+    let step = EXPECTED_SHARE / max_gain;
+    let min_throws = (num_samples_f64 * gain * step) as usize;
+    info!(
+        "Expecting a minimum of {} throws for each coin side. Throws of true: {}. Throws of false: {}.",
+        min_throws, count_true, count_false
+    );
+    assert!(count_true > min_throws);
+    assert!(count_false > min_throws);
```

Comment:
What's the probability of this legitimately failing? If that's not negligible, maybe we should either `#[ignore]` this test by default and in CI, or use a fixed random seed.

Also, agreement doesn't depend on the coin being totally fair. If it's a 40%-60% coin, it would probably still work (almost) just as well. (But I agree that it makes sense to have this test in general. Anything other than 50%-50% would be worth investigating.)

Reply:
These tests with 200 and 50 samples require 42.5% and 31.4% expectations respectively. I set the maximum expectation to 48% if someone tests with 400 samples or more. I agree that this still might fail sometimes and we don't want that. I can tune down the maximum expectation to 40-45% which should be more relaxed.